analyzers/src/SonarAnalyzer.Common/Rules/UseIFormatProviderForParsingDateAndTimeBase.cs

```diff
@@ -0,0 +1,79 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Globalization;
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class UseIFormatProviderForParsingDateAndTimeBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S6580";
+
+    private static readonly string[] parseMethodNames = new[]
+    {
+        nameof(DateTime.Parse),
+        nameof(DateTime.ParseExact),
+        nameof(DateTime.TryParse),
+        nameof(DateTime.TryParseExact)
+    };
+    private static readonly string[] nonDeterministicFormatProviders = new[]
+    {
+        nameof(CultureInfo.CurrentCulture),
+        nameof(CultureInfo.CurrentUICulture),
+        nameof(CultureInfo.DefaultThreadCurrentCulture),
+        nameof(CultureInfo.DefaultThreadCurrentUICulture)
+    };
+    private static readonly KnownType[] temporalTypes = new[]
+    {
+        KnownType.System_DateOnly,
+        KnownType.System_DateTime,
+        KnownType.System_DateTimeOffset,
+        KnownType.System_TimeOnly,
+        KnownType.System_TimeSpan
+    };
+
+    protected override string MessageFormat => "Pass an 'IFormatProvider' to the '{0}' method.";
+
+    protected UseIFormatProviderForParsingDateAndTimeBase() : base(DiagnosticId) { }
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(Language.GeneratedCodeRecognizer, c =>
+        {
+            if (Language.Syntax.NodeIdentifier(c.Node) is { IsMissing: false } identifier
```

Comment:
Please take a look at the [MemberDescriptor](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/MemberDescriptor.cs) class. I think you could use it here, to match the type and the method.

Reply:
I tried. But this rule deals with 20 different methods (5 class x 4 methods), so it's easier for me to do it like this, rather than creating 20 constants or doing a projection with LINQ to get the list of `MemberDescriptor`s